src/main/scala/nl/knaw/dans/easy/multideposit/Main.scala

```diff
@@ -16,188 +16,89 @@
 package nl.knaw.dans.easy.multideposit
 
 import nl.knaw.dans.easy.multideposit.actions._
-import nl.knaw.dans.easy.multideposit.{CommandLineOptions => cmd, MultiDepositParser => parser}
-import org.slf4j.LoggerFactory
-import rx.lang.scala.{Observable, ObservableExtensions}
-import rx.schedulers.Schedulers
+import nl.knaw.dans.lib.logging.DebugEnhancedLogging
 
-import scala.collection.mutable
 import scala.collection.mutable.ListBuffer
-import scala.util.{Failure, Success, Try}
-
-object Main {
-
-  implicit val log = LoggerFactory.getLogger(getClass)
-
-  def main(args: Array[String]) {
-    log.debug("Starting application.")
-    implicit val settings = cmd.parse(args)
-    getActionsStream
-      .doOnError(e => log.error(e.getMessage))
-      .doOnError(e => log.debug(e.getMessage, e))
-      .doOnCompleted { log.info("Finished successfully!") }
-      .doOnTerminate {
-        // close LDAP at the end of the main
-        log.debug("closing ldap")
-        settings.ldap.close()
-      }
-      .subscribe
+import scala.util.Try
 
-    Schedulers.shutdown()
-  }
+object Main extends DebugEnhancedLogging {
 
-  def getActionsStream(implicit settings: Settings): Observable[Action] = {
-    parser.parse(multiDepositInstructionsFile(settings))
-      .flatMap(_.toObservable)
-      .getActions
-      .checkActionPreconditions
-      .runActions
-  }
+  def main(args: Array[String]): Unit = {
+    debug("Starting application")
+    implicit val settings = CommandLineOptions.parse(args)
 
-  // Extension methods are used here to do composition
-  // (see http://reactivex.io/rxscala/comparison.html)
-  implicit class ActionExtensionMethods(val actions: Observable[Action]) extends AnyVal {
-    def checkActionPreconditions = Main.checkActionPreconditions(actions)
+    run
+      .ifFailure { case e =>
+        logger.error(e.getMessage)
+        logger.debug(e.getMessage, e)
+      }
+      .ifSuccess(_ => logger.info("Finished successfully!"))
 
-    def runActions = Main.runActions(actions)
+    logger.debug("closing ldap")
+    settings.ldap.close()
   }
 
-  implicit class DatasetsExtensionMethods(val datasets: Observable[(DatasetID, Dataset)]) extends AnyVal {
-    def getActions(implicit s: Settings) = Main.getActions(datasets)
+  def run(implicit settings: Settings): Try[Unit] = {
+    MultiDepositParser.parse(multiDepositInstructionsFile)
+      .flatMap(getActions(_).reduce(_ compose _).run)
```

Comment:
it did, but only because the parser calls `xs.head`, which throws a `NoSuchElementException` if `xs` is empty. I changed that by replacing the header validation with:

```scala
content.headOption.map(validateDatasetHeaders).getOrElse(Failure(EmptyInstructionsFileException(file)))
```